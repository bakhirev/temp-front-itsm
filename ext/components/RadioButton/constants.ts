export type RadioButtonSize = 'big' | 'small';

export const RADIOBUTTON_SIZE_SMALL = '18px';
export const RADIOBUTTON_SIZE_BIG = '20px';
export const RADIOBUTTON_MARGIN_SMALL = '1px';
export const RADIOBUTTON_MARGIN_BIG = '2px';
export const RADIOBUTTON_MIN_HEIGHT_SMALL = '20px';
export const RADIOBUTTON_MIN_HEIGT_BIG = '24px';
export const LABEL_PADDING = '2px 2px 2px 8px';
export const OUTER_CIRCLE_OFFSET_SMALL = '7px';
export const OUTER_CIRCLE_OFFSET_BIG = '8px';
