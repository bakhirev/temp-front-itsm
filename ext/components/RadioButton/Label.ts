import styled, { css } from 'styled-components';

import { Body2Short, Body1Short } from '../Typography';
import { DEFAULT_THEME } from '../common';
import type { ITheme } from '../themes';

import { LABEL_PADDING } from './constants';
import type { RadioButtonSize } from './constants';

type LabelProps = {
  disabled?: boolean;
};

const commonLabelStyles = (theme: ITheme, disabled?: boolean) => css`
  padding: ${LABEL_PADDING};
  color: ${disabled ? theme.color.neutral[30] : theme.color.neutral[90]};
  user-select: none;
  &:focus {
    outline: none;
  }
`;

const SmallLabel = styled(Body2Short).attrs({ tabIndex: -1 })<LabelProps>`
  ${({ theme, disabled }) => commonLabelStyles(theme, disabled)}
`;

SmallLabel.defaultProps = {
  theme: DEFAULT_THEME,
};

const BigLabel = styled(Body1Short).attrs({ tabIndex: -1 })<LabelProps>`
  ${({ theme, disabled }) => commonLabelStyles(theme, disabled)}
`;

BigLabel.defaultProps = {
  theme: DEFAULT_THEME,
};

export const Label = (size: RadioButtonSize) => (size === 'small' ? SmallLabel : BigLabel);
