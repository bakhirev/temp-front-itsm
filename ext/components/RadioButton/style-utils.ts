import { css } from 'styled-components';

import type { ITheme } from '../themes';

import { RadioButtonSize, OUTER_CIRCLE_OFFSET_BIG, OUTER_CIRCLE_OFFSET_SMALL } from './constants';

export const RadioScaleAnimation = css`
  @keyframes RadioScale {
    0% {
      transform: translate(-50%, -50%) scale(0);
    }
    80% {
      transform: translate(-50%, -50%) scale(1.2);
    }
    100% {
      transform: translate(-50%, -50%) scale(1);
    }
  }
`;

export const MarkBeforeContainer = (size: RadioButtonSize) => css`
  top: -${size === 'small' ? OUTER_CIRCLE_OFFSET_SMALL : OUTER_CIRCLE_OFFSET_BIG};
  left: -${size === 'small' ? OUTER_CIRCLE_OFFSET_SMALL : OUTER_CIRCLE_OFFSET_BIG};
  bottom: -${size === 'small' ? OUTER_CIRCLE_OFFSET_SMALL : OUTER_CIRCLE_OFFSET_BIG};
  right: -${size === 'small' ? OUTER_CIRCLE_OFFSET_SMALL : OUTER_CIRCLE_OFFSET_BIG};
`;

export const commonRadioButtonComponentStyle = (
  theme: ITheme,
  type: 'Hover' | 'Pressed' | 'Focus',
  disabled?: boolean
) => css`
  input + div {
    &:before {
      background-color: ${disabled ? 'transparent' : theme.color.opacity['black' + type]};
    }
  }
`;
