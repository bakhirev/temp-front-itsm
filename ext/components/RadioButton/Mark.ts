import styled, { css } from 'styled-components';
import { size as Size } from 'polished';

import { DEFAULT_THEME } from '../common';

import {
  RADIOBUTTON_MARGIN_BIG,
  RADIOBUTTON_MARGIN_SMALL,
  RADIOBUTTON_SIZE_BIG,
  RADIOBUTTON_SIZE_SMALL,
} from './constants';
import { MarkBeforeContainer } from './style-utils';
import type { RadioButtonSize } from './constants';

type MarkProps = {
  size: RadioButtonSize;
  disabled?: boolean;
};

export const Mark = styled.div<MarkProps>`
  border-radius: 50%;
  display: flex;
  flex: 0 0 auto;
  position: relative;
  ${({ size, theme, disabled }) => css`
    background-color: ${theme.color.neutral.white};
    border: 1px solid ${disabled ? theme.color.neutral[30] : theme.color.neutral[50]};
    margin: ${size === 'small' ? RADIOBUTTON_MARGIN_SMALL : RADIOBUTTON_MARGIN_BIG};
    ${size === 'small' ? Size(RADIOBUTTON_SIZE_SMALL) : Size(RADIOBUTTON_SIZE_BIG)};
  `}

  &:focus {
    outline: none;
  }

  &:before {
    background-color: transparent;
    border-radius: 50%;
    content: '';
    position: absolute;
    ${({ size }) => MarkBeforeContainer(size)};
  }

  &:after {
    background: transparent;
    border-radius: 50%;
    border: 0 solid transparent;
    content: '';
    left: 50%;
    position: absolute;
    top: 50%;
    transform: translate(-50%, -50%);
    transition: all 0.2s;
    ${Size('100%')}
  }
`;

Mark.defaultProps = {
  theme: DEFAULT_THEME,
};
