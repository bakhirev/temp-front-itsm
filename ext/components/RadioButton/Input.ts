import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../common';

import { Mark } from './Mark';
import { RadioScaleAnimation } from './style-utils';

export const Input = styled.input<{ disabled?: boolean }>`
  height: 0;
  position: absolute;
  width: 0;
  ${RadioScaleAnimation}

  &:focus {
    outline: none;
  }

  &:checked ~ ${Mark} {
    ${({ theme, disabled }) => css`
      border: 1px solid ${disabled ? theme.color.neutral[30] : theme.color.primary[60]};
    `}

    &:after {
      ${({ theme, disabled }) => css`
        border: 3px solid ${disabled ? theme.color.neutral[30] : theme.color.primary[60]};
      `};
    }
  }
`;

Input.defaultProps = {
  theme: DEFAULT_THEME,
};
