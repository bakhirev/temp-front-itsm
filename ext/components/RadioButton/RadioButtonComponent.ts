import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../common';

import { Input } from './Input';
import { Mark } from './Mark';
import type { RadioButtonSize } from './constants';
import { RADIOBUTTON_MIN_HEIGHT_SMALL, RADIOBUTTON_MIN_HEIGT_BIG } from './constants';
import { commonRadioButtonComponentStyle } from './style-utils';

type RadioButtonComponentProps = {
  size: RadioButtonSize;
  error?: boolean;
  disabled?: boolean;
};

export const RadioButtonComponent = styled.div<RadioButtonComponentProps>`
  box-sizing: border-box;
  display: inline-flex;
  align-items: flex-start;
  min-height: ${({ size }) =>
    size === 'small' ? RADIOBUTTON_MIN_HEIGHT_SMALL : RADIOBUTTON_MIN_HEIGT_BIG};

  ${({ error, theme }) =>
    error &&
    css`
      ${Mark} {
        border: 1px solid ${theme.color.error[60]};
      }
    `}

  ${({ theme, disabled }) => css`
    cursor: ${disabled ? 'default' : 'pointer'};
    ${Input}:checked ~ ${Mark} {
      border-color: ${disabled ? theme.color.neutral[30] : theme.color.primary[60]};
    }
    ${Mark} {
      border-color: ${disabled ? theme.color.neutral[30] : theme.color.neutral[50]};
    }
    &:hover {
      ${commonRadioButtonComponentStyle(theme, 'Hover', disabled)}
    }
    &:focus {
      outline: none;
      ${commonRadioButtonComponentStyle(theme, 'Focus', disabled)}
    }
    &:active {
      ${commonRadioButtonComponentStyle(theme, 'Pressed', disabled)}
    }
  `}
`;

RadioButtonComponent.defaultProps = {
  theme: DEFAULT_THEME,
};
