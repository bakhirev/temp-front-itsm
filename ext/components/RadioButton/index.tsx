import React, { FC, MouseEvent, KeyboardEvent } from 'react';

import { KEY_CODES } from '../common';

import { Mark } from './Mark';
import { Input } from './Input';
import { Label } from './Label';
import { RadioButtonComponent } from './RadioButtonComponent';
import { RadioButtonSize } from './constants';

export interface IRadioButtonProps {
  /** Коллбэк на изменение состояния. Срабатывает при клике или нажатии Enter/Space при переходе по табу */
  onChange: (event: MouseEvent<HTMLDivElement> | KeyboardEvent<HTMLDivElement>, value: any) => void;
  /** Подпись к радиокнопке */
  label: string;
  /** Значение радиокнопки */
  value: any;
  /** Состояние радиокнопки: выбрана/не выбрана */
  checked: boolean;
  /** Размер радиокнопки */
  size?: RadioButtonSize;
  /** Отключение радиокнопки */
  disabled?: boolean;
  /** Состояние ошибки радиокнопки */
  error?: boolean;
  /** Атрибут name радиокнопки */
  name?: string;
  /** Имя класса для переопределения стилей */
  className?: string;
  /** data-test-id атрибут для тестирования компонента */
  dataTestId?: string;
}

export const RadioButton: FC<IRadioButtonProps> = ({
  label,
  name,
  value,
  checked,
  disabled,
  error,
  size = 'big',
  className,
  dataTestId,
  onChange,
}: IRadioButtonProps) => {
  const handleClick = (event: MouseEvent<HTMLDivElement>) => {
    if (!disabled) {
      onChange && onChange(event, value);
    }
  };

  const handleKeyUp = (event: KeyboardEvent<HTMLDivElement>) => {
    if (!disabled && (event.keyCode === KEY_CODES.ENTER || event.keyCode === KEY_CODES.SPACEBAR)) {
      onChange && onChange(event, value);
    }
  };
  const LabelWrapper = Label(size);
  return (
    <RadioButtonComponent
      tabIndex={1}
      size={size}
      error={!checked && error}
      disabled={disabled}
      onKeyUp={handleKeyUp}
      onClick={handleClick}
      className={className}
      data-test-id={dataTestId}
    >
      <Input
        type="radio"
        readOnly
        name={name}
        checked={checked}
        value={value || ''}
        disabled={disabled}
      />
      <Mark tabIndex={-1} size={size} disabled={disabled} />
      <LabelWrapper disabled={disabled}>{label}</LabelWrapper>
    </RadioButtonComponent>
  );
};
