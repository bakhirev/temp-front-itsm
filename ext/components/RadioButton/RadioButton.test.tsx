import React from 'react';
import { shallow } from 'enzyme';

import { RadioButton } from '../RadioButton';

jest.mock('../common', () => ({
  DEFAULT_THEME: {},
}));

describe('RadioButton', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const radiobuttonRequiredProps = {
    onChange: jest.fn(),
    label: 'RadioButton',
    value: 'radio1',
    checked: false,
  };

  it('should render component with required properties', () => {
    const wrapper = shallow(<RadioButton {...radiobuttonRequiredProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with disabled state', () => {
    const wrapper = shallow(<RadioButton {...radiobuttonRequiredProps} disabled />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with small size', () => {
    const wrapper = shallow(<RadioButton {...radiobuttonRequiredProps} size="small" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with error state', () => {
    const wrapper = shallow(<RadioButton {...radiobuttonRequiredProps} error />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should call onChange when user clicks on component', () => {
    const wrapper = shallow(<RadioButton {...radiobuttonRequiredProps} />);
    wrapper.simulate('click');
    const { onChange } = radiobuttonRequiredProps;
    expect(onChange).toBeCalledTimes(1);
  });
});
