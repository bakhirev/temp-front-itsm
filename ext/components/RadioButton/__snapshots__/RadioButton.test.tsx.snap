// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton should render component with disabled state 1`] = `
<styled.div
  disabled={true}
  onClick={[Function]}
  onKeyUp={[Function]}
  size="big"
  tabIndex={1}
  theme={Object {}}
>
  <styled.input
    checked={false}
    disabled={true}
    readOnly={true}
    theme={Object {}}
    type="radio"
    value="radio1"
  />
  <styled.div
    disabled={true}
    size="big"
    tabIndex={-1}
    theme={Object {}}
  />
  <Styled(styled.div)
    disabled={true}
    theme={Object {}}
  >
    RadioButton
  </Styled(styled.div)>
</styled.div>
`;

exports[`RadioButton should render component with error state 1`] = `
<styled.div
  error={true}
  onClick={[Function]}
  onKeyUp={[Function]}
  size="big"
  tabIndex={1}
  theme={Object {}}
>
  <styled.input
    checked={false}
    readOnly={true}
    theme={Object {}}
    type="radio"
    value="radio1"
  />
  <styled.div
    size="big"
    tabIndex={-1}
    theme={Object {}}
  />
  <Styled(styled.div)
    theme={Object {}}
  >
    RadioButton
  </Styled(styled.div)>
</styled.div>
`;

exports[`RadioButton should render component with required properties 1`] = `
<styled.div
  onClick={[Function]}
  onKeyUp={[Function]}
  size="big"
  tabIndex={1}
  theme={Object {}}
>
  <styled.input
    checked={false}
    readOnly={true}
    theme={Object {}}
    type="radio"
    value="radio1"
  />
  <styled.div
    size="big"
    tabIndex={-1}
    theme={Object {}}
  />
  <Styled(styled.div)
    theme={Object {}}
  >
    RadioButton
  </Styled(styled.div)>
</styled.div>
`;

exports[`RadioButton should render component with small size 1`] = `
<styled.div
  onClick={[Function]}
  onKeyUp={[Function]}
  size="small"
  tabIndex={1}
  theme={Object {}}
>
  <styled.input
    checked={false}
    readOnly={true}
    theme={Object {}}
    type="radio"
    value="radio1"
  />
  <styled.div
    size="small"
    tabIndex={-1}
    theme={Object {}}
  />
  <Styled(styled.div)
    theme={Object {}}
  >
    RadioButton
  </Styled(styled.div)>
</styled.div>
`;
